librz/core/cmd/cmd_print.c

```diff
@@ -6563,6 +6563,238 @@ RZ_IPI RzCmdStatus rz_print_utf32be_handler(RzCore *core, int argc, const char *
+	RZ_CORE_PRINT_FORMAT_TYPE_OCTAL = 0,
+	RZ_CORE_PRINT_FORMAT_TYPE_INTEGER,
+	RZ_CORE_PRINT_FORMAT_TYPE_HEXADECIMAL,
+	RZ_CORE_PRINT_FORMAT_TYPE_INVALID,
+} RzCorePrintFormatType;
+
+static inline st8 format_type_to_base(const RzCorePrintFormatType format, const ut8 n) {
+	static const st8 bases[][9] = {
+		{ 0, 8 },
+		{ 0, -1, -10, [4] = 10, [8] = -8 },
+		{ 0, 16, 32, [4] = 32, [8] = 64 },
+	};
+	if (format >= RZ_CORE_PRINT_FORMAT_TYPE_INVALID || n >= sizeof(bases[0])) {
+		return 0;
+	}
+	return bases[format][n];
+}
+
+static inline void fix_size_from_format(const RzCorePrintFormatType format, ut8 *size) {
+	if (format != RZ_CORE_PRINT_FORMAT_TYPE_INTEGER) {
+		return;
+	}
+	static const st8 sizes[] = {
+		0, 4, 2, [4] = 4, [8] = 4
+	};
+	if (*size >= sizeof(sizes)) {
+		return;
+	}
+	*size = sizes[*size];
+}
+
+static inline void len_fixup(RzCore *core, ut64 *addr, int *len) {
+	if (!len || *len >= 0) {
+		return;
+	}
+	*len = -*len;
+	if (*len > core->blocksize_max) {
+		RZ_LOG_ERROR("this block size is too big (0x%" PFMT32x
+			     " < 0x%" PFMT32x ").",
+			*len,
+			core->blocksize_max);
+		*len = (int)core->blocksize_max;
+	}
+	*addr = *addr - *len;
+}
+
+static inline bool print_dump(RzCore *core, const RzCmdStateOutput *state, ut64 addr, ut8 n, int len, const RzCorePrintFormatType format) {
+	st8 base = format_type_to_base(format, n);
+	if (!base) {
+		return false;
+	}
+	len_fixup(core, &addr, &len);
+	ut8 *buffer = malloc(len);
+	if (!buffer) {
+		return false;
+	}
+	rz_io_read_at(core->io, addr, buffer, len);
+
+	RZ_LOG_VERBOSE("Dump 0x%" PFMT64x " %d %d %d\n", addr, n, len, base);
+
+	rz_print_init_rowoffsets(core->print);
+	//	bool show_offset = rz_config_get_i(core->config, "hex.offset");
+	//	if (show_offset) {
+	//		core->print->flags |= RZ_PRINT_FLAGS_OFFSET;
+	//	} else {
+	//		core->print->flags &= ~RZ_PRINT_FLAGS_OFFSET;
+	//	}
+	//	bool hex_header = rz_config_get_i(core->config, "hex.header");
+	//	if (hex_header) {
+	//		core->print->flags |= RZ_PRINT_FLAGS_HEADER;
+	//	} else {
+	//		core->print->flags &= ~RZ_PRINT_FLAGS_HEADER;
+	//	}
+	//	/* Don't show comments in default case */
+	core->print->use_comments = false;
+	RzOutputMode mode = state ? state->mode : RZ_OUTPUT_MODE_STANDARD;
+
+	switch (mode) {
+	case RZ_OUTPUT_MODE_JSON:
+		rz_print_jsondump(core->print, buffer, len, n * 8);
+		break;
+	case RZ_OUTPUT_MODE_STANDARD:
+		fix_size_from_format(format, &n);
+		rz_print_hexdump(core->print, addr,
+			buffer, len, base, (int)n, 1);
+		break;
+	default:
+		rz_warn_if_reached();
+		free(buffer);
+		return false;
+	}
+	free(buffer);
+	return true;
+}
+
+RZ_IPI RzCmdStatus rz_print_hexdump_signed_int_handler(RzCore *core, int argc, const char **argv, RzCmdStateOutput *state) {
+	int n = (int)rz_num_math(core->num, argv[1]);
+	int len = argc > 2 ? (int)rz_num_math(core->num, argv[2]) : (int)core->blocksize;
+	return bool2status(print_dump(core, state, core->offset, n, len, RZ_CORE_PRINT_FORMAT_TYPE_INTEGER));
+}
+
```

Comment:
Some of these functions could be converted into returning `char *` so that they would also be usable as an API, then you could also move them to [`librz/core/cprint.c`](https://github.com/rizinorg/rizin/blob/dev/librz/core/cprint.c).

Reply:
But I feel that most of the existing APIs use `RzCons` to output, but it is not reasonable to change them to return `char*` one by one.

So I guess it might be reasonable to use something like this
https://github.com/rizinorg/cutter/blob/1dbad3d1edf3a2061de7fbe6e01af7d3da8494c8/src/dialogs/XrefsDialog.cpp#L134-L145

We can continue to use these APIs using `RzCons` instead of `rz_core_cmd*`

Reply:
@imbillow, I don't think it's a good approach, more like a temporary solution. Also, Cutter isn't a sole user of the API, we wait to become a generally-usable library, even with various bindings; see https://github.com/rizinorg/rz-bindgen
I don't we have another choice than to return C structures wherever possible/make sense or strings where it doesn't.